Execute explicit instantiations of compute_intersection_of_cells()

jh66637 opened the pull request:
When calling ```compute_intersection_of_cells(vetices,...)``` deal.ii I always end up in the function of the header. Removing the definition from the header and adding the overloads which can be called from ```compute_intersection_of_cells(iterator,...)``` due to its explicit instantiations resolves this issue. 

Since intersections between vertices do not make any sense, I restricted the explicit template instantiations here. Intersections of two quads and a quad in a line with spacedim 3 are however conceivable. I think implementation of this functionality should be done in a separate PR.

Note that all intersections are tested through ```compute_intersection_of_cells(iterator,...)```. The additional test is only meant to ensure the there are no linker errors if a declaration is missing.

@fdrmrc

drwells commented:
I'm not sure that this is going to permanently fix the problem - even if it works, it is relying on some very precise rules regarding template specializations followed by template instantiations.

I think that we agreed at the point of the last release that these APIs were not totally stable. Would you be OK with doing something like

```cpp
  template <int dim0, int dim1, int spacedim>
  std::vector<std::array<Point<spacedim>, dim1 + 1>>
  compute_intersection_of_cells(
    const ArrayView<Point<spacedim>> &vertices0,
    const ArrayView<Point<spacedim>> &vertices1,
    const double                      tol = 1e-9);
```

and then making the intersection behavior depend on the run-time length of `vertices0` and `vertices1` instead? That would, I believe, allow us to combine all of the instantiations into one function (and of course we can use `if constexpr()` in it as needed).

drwells commented on include/deal.II/cgal/intersections.h:
```diff
@@ -61,20 +61,15 @@ namespace CGALWrappers
 
   /**
    * Same function as above, but working directly with vertices.
+   * @note the vertices have to be given in CGAL order
```
If you could document what `dim0` etc. do that would help a lot too.

jh66637 replied:
Sure, I can do that.

jh66637 commented:
In general, I am fine with changing the interface in this way. I just have to look into the implementation.
@fdrmrc Any objections from your side?

fdrmrc commented:
That works for me too.

bangerth commented:
Any news here?

jh66637 commented:
From my point of view, this is ready. Thanks to @fdrmrc for the big help :)
Currently, only the test "cgal/cgal_triangulation_05.release" fails on my machine. I don't think this is related to the PR, since the functions that are called in the test are not touched.

masterleinad commented on include/deal.II/cgal/utilities.h:
```diff
@@ -601,28 +601,24 @@ namespace CGALWrappers
     const typename dealii::Triangulation<dim, spacedim>::cell_iterator &cell,
-    const Mapping<dim, spacedim> &                                      mapping)
+    const Mapping<dim, spacedim> &                                      mapping,
+    std::vector<Point<spacedim>> &                                      dst)
```
You should also assert that `dst` has size `n_vertices`. Is there a reason not to return the result as it was previously done?

jh66637 replied:
This change was done by @fdrmrc. From my side, both is fine.

fdrmrc replied:
No, there's no particular reason actually. I've just realized that if you want to use this version we need to update the documentation for this function. Sorry about that.

masterleinad replied:
My preference would be to return the result again but at the very least the documentation must match what the function is doing.

fdrmrc replied:
So done

jh66637 commented:
I added the requested Assert and squashed the PR.

bangerth commented on source/cgal/intersections.cc:
```diff
@@ -359,400 +355,449 @@ namespace CGALWrappers
+
+  template <int dim0, int dim1, int spacedim>
+  std::vector<std::array<Point<spacedim>, dim1 + 1>>
+  compute_intersection_of_cells(const ArrayView<Point<spacedim>> &vertices0,
+                                const ArrayView<Point<spacedim>> &vertices1,
```
assertion

fdrmrc replied:
This function will call, depending on the runtime length of the vertices, the previous functions computing intersections (where all the assertions have been added). But I don't see which assumption I should make on the number of vertices for this particular function. I could require that both vertices have size>0, but I'm not sure if you're referring to that.

bangerth replied:
I see now. I just marked this place up like all of the others where an `ArrayView` is passed, assuming that there is an assumed-size for these arrays. Ignore the comment here then.

bangerth reviewed (commented):
Here's another review. It's really just the same thing many times over, nothing of substane that should take you more than ten or twenty minutes.

fdrmrc commented on source/cgal/intersections.cc:
```diff
@@ -359,400 +355,449 @@ namespace CGALWrappers
+              if (const CGALSegment2 *s =
+                    boost::get<CGALSegment2>(&*intersection))
+                {
+                  vertices.push_back(
+                    {{CGALWrappers::cgal_point_to_dealii_point<2>((*s)[0]),
+                      CGALWrappers::cgal_point_to_dealii_point<2>((*s)[1])}});
```
Unfortunately no. `std::array<>` has no user-defined constructor

bangerth replied:
Ah, I see. In that case just ignore these comments everywhere.

fdrmrc commented on source/cgal/intersections.cc:
```diff
@@ -359,400 +355,449 @@ namespace CGALWrappers
+          for (const auto &c1 : tria1.finite_cell_handles())
+            {
+              const auto &                 tet1 = tria1.tetrahedron(c1);
+              [[maybe_unused]] const auto &tetg1 =
```
Yes, CGAL requires C++17, but you're totally right about the `[[maybe_unused]]`. I removed it, thanks.

jh66637 replied:
Removing it leads to ugly compiler warnings in my case. To suppress this, I added (void)tetg1.

fdrmrc commented on source/cgal/intersections.cc:
```diff
@@ -225,82 +225,80 @@ namespace CGALWrappers
                                            CGALTriangle2,
                                            std::vector<CGALPoint2>>>
-    compute_intersection(const std::array<Point<2>, 3> &first_simplex,
-                         const std::array<Point<2>, 3> &second_simplex)
+    compute_intersection_triangle_triangle(
+      const ArrayView<Point<spacedim>> &triangle0,
+      const ArrayView<Point<spacedim>> &triangle1)
     {
```
Yes, these internal functions are called depending on the runtime length of the vertices in `compute_intersections_of_cells(),` so the precondition will not be violated in practice. But as a precondition an assertion should definitely be there, thanks.

fdrmrc commented on source/cgal/intersections.cc:
```diff
@@ -316,26 +314,24 @@ namespace CGALWrappers
-    compute_intersection(const std::array<Point<3>, 2> &first_simplex,
-                         const std::array<Point<3>, 4> &second_simplex)
+    compute_intersection_tetra_segment(
+      const std::array<Point<3>, 2> &tetrahedron,
+      const std::array<Point<3>, 4> &segment)
```
You're right. Fixed

fdrmrc commented on source/cgal/intersections.cc:
```diff
@@ -316,26 +314,24 @@ namespace CGALWrappers
-        second_simplex.end(),
-        pts1.begin(),
-        [&](const Point<3> &p) {
+        segment.begin(), segment.end(), pts1.begin(), [&](const Point<3> &p) {
           return CGALWrappers::dealii_point_to_cgal_point<CGALPoint3>(p);
         });
```
This will not compile because the template parameter `UnaryOperation` from [std::transform](https://en.cppreference.com/w/cpp/algorithm/transform) will not be deduced.

bangerth replied:
No, there should be no problem. The `UnaryType` is simply the type of the pointer-to-function `&CGALWrappers::dealii_point_to_cgal_point<CGALPoint3>`. Or am I missing something?

jh66637 replied:
The return type of the function pointer is the function signature which includes ```Point<dim>``` (which the compiler can not deduce). Thus, we need ```&CGALWrappers::dealii_point_to_cgal_point<CGALPoint3,dim>```

bangerth commented:
/rebuild

jh66637 commented:
I think all requested changes have been made. Once again, big thanks to @fdrmrc :)

jh66637 commented:
@bangerth @drwells @masterleinad  
Is there anything left to do to get this merged? :)

drwells commented on source/cgal/intersections.cc:
```diff
@@ -225,130 +225,133 @@ namespace CGALWrappers
     // signatures follow the convection that the first entity has an intrinsic
     // dimension higher than the second one.
 
+    template <int spacedim>
```
it looks like this is only for spacedim = 2. Do we need the template parameter?

jh66637 replied:
I deleted the obsolete template parameter for both cases.

drwells commented on source/cgal/intersections.cc:
```diff
@@ -225,130 +225,133 @@ namespace CGALWrappers
+    }
 
 
+    template <int spacedim>
```
Same here

masterleinad reviewed (approved):
Looks OK to me.

jh66637 commented:
@bangerth I think I addressed all comments. If you approve the changes, I will squash and proceed with the missing parts to collect CGAL intersections using MPI.

jh66637 commented:
@bangerth can we proceed here?

peterrum commented:
/rebuild

peterrum commented:
@jh66637 Could you rebase and force push again? On of the tests is not starting!?
